lib/RestructureCFG/InlineDivergentScopesPass.cpp

```diff
@@ -316,42 +385,182 @@ static void performIDS(DivergenceDescriptor &DivergenceDescriptor,
+/// they insist on the same conditional, have non-overlapping `Exit` blocks, and
+/// do not have any overlapping `DivergentSuccessors`
+static bool isCompatible(const SmallVector<DivergenceDescriptor> &Collection,
+                         const DivergenceDescriptor &NewDescriptor) {
+  // Skip checking if collection is empty
+  if (Collection.empty())
+    return true;
+
+  // Check that Conditional matches
+  if (Collection[0].Conditional != NewDescriptor.Conditional)
+    return false;
+
+  // Check for non-overlapping Exit blocks
+  for (const auto &Existing : Collection) {
+    if (Existing.Exit == NewDescriptor.Exit)
+      return false;
+
+    // Check for non-overlapping DivergentSuccessors
+    for (BasicBlock *ExistingSucc : Existing.DivergentSuccessors) {
+      for (BasicBlock *NewSucc : NewDescriptor.DivergentSuccessors) {
+        if (ExistingSucc == NewSucc)
+          return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+/// Helper function which attempts to perform multiple IDS transformations
+static bool
+tryMultipleIDS(const ScopeGraphBuilder &SGBuilder,
+               const SmallVector<DivergenceDescriptor> &DivergenceDescriptors,
+               BasicBlock *PlaceHolderTarget) {
+
+  // We attempt to perform multiple IDS transformations all at once
+  // We start by collecting all the `Conditional`s involved in at least one
+  // divergence
+  SmallSetVector<BasicBlock *, 4> Conditionals;
+  for (auto DivergenceDescriptor : DivergenceDescriptors) {
+    Conditionals.insert(DivergenceDescriptor.Conditional);
+  }
+
+  // For each conditional, we try to collect all the multiple divergences that
+  // involves such `Conditional`
+  for (BasicBlock *Conditional : Conditionals) {
+
+    // We accumulate in `Result` all the `DivergenceDescriptor`s compatible for
+    // a certain `Conditional`
+    SmallVector<DivergenceDescriptor> CompatibleDivergenceDescriptors;
+    for (auto DivergenceDescriptor : DivergenceDescriptors) {
+
+      // We manually initialize `CompatibleDivergenceDescriptors` when we
+      // encounter the first `DivergenceDescriptor` involved for the
+      // `Conditional` under analysis
+      if (CompatibleDivergenceDescriptors.empty()) {
+        if (DivergenceDescriptor.Conditional == Conditional) {
+          revng_assert(isCompatible(CompatibleDivergenceDescriptors,
+                                    DivergenceDescriptor));
+
+          // Initialize the multiple divergence set
+          CompatibleDivergenceDescriptors.push_back(DivergenceDescriptor);
+        }
+      } else {
+
+        // If we find a `DivergenceDescriptor` compatible with the current
+        // `Conditional`, we add it to the pool of candidates
+        if (isCompatible(CompatibleDivergenceDescriptors,
+                         DivergenceDescriptor)) {
+          CompatibleDivergenceDescriptors.push_back(DivergenceDescriptor);
+        }
+      }
+    }
+
+    if (CompatibleDivergenceDescriptors.size() >= 1) {
```

Comment:
This should always be true.

Reply:
I completely removed this `if`, and left this check as an `assert` as a failsafe here https://github.com/revng/revng/blob/1b4cabba417d846b3f4e47ca33b3f6b7b0a15234/lib/RestructureCFG/InlineDivergentScopesPass.cpp#L525